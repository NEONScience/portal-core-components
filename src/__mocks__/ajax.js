/**
   Ajax Mock

   Tests for any components using rjsx/ajax.getJSON can use this to inject a synchronous response.

   Usage:

   import { mockAjaxResponse, mockAjaxError } from 'path/to/src/__mocks__/ajax';

   // Set a mock response (for ajax() AND ajax.getJSON() methods)
   mockAjaxResponse({ ...response });

   // Set a mock error (for ajax() AND ajax.getJSON() methods)
   mockAjaxError('fail');
*/

import { of, throwError } from 'rxjs';

import { ajax } from 'rxjs/ajax';

jest.mock('rxjs/ajax');

ajax.mockReturnValue(of({}));

ajax.getJSON = jest.fn();
ajax.getJSON.mockReturnValue(of({}));

export function mockRawAjaxResponse(response = {}) {
  ajax.mockReset();
  ajax.mockReturnValue(of(response));
}

export function mockRawAjaxError(error = '') {
  ajax.mockReset();
  ajax.mockImplementation(() => throwError(new Error(error)));
}

export function mockGetJSONAjaxResponse(response = {}) {
  ajax.getJSON.mockReset();
  ajax.getJSON.mockReturnValue(of(response));
}

export function mockGetJSONAjaxError(error = '') {
  ajax.getJSON.mockReset();
  ajax.getJSON.mockImplementation(() => throwError(new Error(error)));
}

export function mockAjaxResponse(response = {}) {
  mockRawAjaxResponse(response);
  mockGetJSONAjaxResponse(response);
}

export function mockAjaxError(error = '') {
  mockRawAjaxError(error);
  mockGetJSONAjaxError(error);
}
