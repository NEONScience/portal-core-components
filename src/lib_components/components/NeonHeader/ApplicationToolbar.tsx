import React from 'react';
import IconButton from '@material-ui/core/IconButton';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Grow from '@material-ui/core/Grow';
import Paper from '@material-ui/core/Paper';
import Popper from '@material-ui/core/Popper';
import MenuItem from '@material-ui/core/MenuItem';
import MenuList from '@material-ui/core/MenuList';
import Tooltip from '@material-ui/core/Tooltip';
import AppsIcon from '@material-ui/icons/Apps';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

const applications = {
  token: 'token',
  data: {
    apps: [
      {
        name: 'NEON Single Sign-On',
        url: 'https://data-neonscience.auth0.com/login?state=hKFo2SBiSjg3dklhMnV6bXFzaUtFQ3dFTl9sdEM4TWI2T0h2N6FupWxvZ2luo3RpZNkgZ2I5Uk1TZm1ZZ0NuSVNYd2JOSUVCQjR1UEVYZktZaTWjY2lk2SBPbjdYRkdrZTROanRZQU03bnE1MUxCcFNFakdGNndtMg&client=On7XFGke4NjtYAM7nq51LBpSEjGF6wm2&protocol=oauth2&redirect_uri=https%3A%2F%2Fdata.neonscience.org%2Fauth0%2Fcallback&audience=https%3A%2F%2Fdata-neonscience.auth0.com%2Fapi%2Fv2%2F&scope=openid%20profile%20email%20read%3Acurrent_user%20create%3Acurrent_user_metadata%20update%3Acurrent_user_metadata&response_type=code',
      },
      {
        name: 'Data Explorer',
        url: 'https://data.neonscience.org/data-products/explore',
      },
      {
        name: 'Application 3',
        url: 'https://example.com/app3',
      },
    ],
  },
};

const useStyles = makeStyles((theme: Theme) => createStyles({
  paper: {
    marginRight: theme.spacing(2),
  },
  toolbarContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    width: '100%',
    boxShadow: '0px 0px 0px rgb(0 0 0 / 0%), 0px 1px 3px rgb(0 0 0 / 35%)',
  },
  toolbarButtons: {
    display: 'flex',
    marginLeft: 'auto',
    marginRight: theme.spacing(2.5),
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
  },
  appBar: {
    background: 'white',
    borderTop: '2px solid transparent',
    marginTop: '-8px',
  },
  arrow: {
    color: theme.palette.common.black,
  },
  tooltip: {
    backgroundColor: theme.palette.common.black,
    color: 'white',
  },
  appButton: {
    marginLeft: theme.spacing(1),
  },
  horizontalMenu: {
    '> div': {
      display: 'inline-block',
    },
  },
}));

export default function ApplicationToolbar() {
  const classes = useStyles();
  const [open, setOpen] = React.useState(false);
  const anchorRef = React.useRef<HTMLButtonElement>(null);
  const [selectedIndex, setSelectedIndex] = React.useState(1);
  const { apps } = applications.data;

  const handleToggle = () => {
    setOpen((prevOpen) => !prevOpen);
  };

  const handleClose = (event: React.MouseEvent<EventTarget>) => {
    if (anchorRef.current && anchorRef.current.contains(event.target as HTMLElement)) {
      return;
    }
    setOpen(false);
  };

  function handleListKeyDown(event: React.KeyboardEvent) {
    if (event.key === 'Tab') {
      event.preventDefault();
      setOpen(false);
    }
  }

  const handleMenuItemClick = (event: React.MouseEvent<EventTarget>, index: number) => {
    setSelectedIndex(index);
    const app = applications.data.apps[index];
    window.location.href = app.url;
  };

  // return focus to the button when we transitioned from !open -> open
  const prevOpen = React.useRef(open);
  React.useEffect(() => {
    if (prevOpen.current === true && open === false) {
      anchorRef.current!.focus();
    }
    prevOpen.current = open;
  }, [open]);

  return (
    <div className={classes.toolbarContainer}>
      <div className={classes.toolbarButtons}>
        <Tooltip
          title="NEON Applications"
          aria-label="NEON Applications"
          arrow
          classes={classes}
        >
          <IconButton
            ref={anchorRef}
            style={{ color: 'black' }}
            aria-label="more"
            aria-controls={open ? 'menu-list-grow' : undefined}
            aria-haspopup="true"
            onClick={handleToggle}
          >
            <AppsIcon />
          </IconButton>
        </Tooltip>

        <Popper open={open} anchorEl={anchorRef.current} role={undefined} transition>
          {({ TransitionProps, placement }) => (
            <Grow
              {...TransitionProps}
              style={{ transformOrigin: placement === 'bottom' ? 'center top' : 'center bottom' }}
              timeout={1000}
            >
              <Paper>
                <ClickAwayListener onClickAway={handleClose}>
                  <MenuList
                    autoFocusItem={open}
                    id="menu-list-grow"
                    onKeyDown={handleListKeyDown}
                    className={classes.horizontalMenu}
                  >
                    <MenuItem onClick={handleClose}>Profile</MenuItem>
                    <MenuItem onClick={handleClose}>My account</MenuItem>
                    <MenuItem onClick={handleClose}>Logout</MenuItem>
                    {apps.map((app, index) => (
                      <MenuItem
                        onClick={(event) => handleMenuItemClick(event, index)}
                        key={app.url}
                        selected={index === selectedIndex}
                      >
                        {app.name}
                      </MenuItem>
                    ))}
                  </MenuList>
                </ClickAwayListener>
              </Paper>
            </Grow>
          )}
        </Popper>
      </div>
    </div>
  );
}
