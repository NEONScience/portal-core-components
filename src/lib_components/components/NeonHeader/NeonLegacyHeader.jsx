import React from 'react';
import PropTypes from 'prop-types';

import { makeStyles } from '@material-ui/core/styles';
import Divider from '@material-ui/core/Divider';

import NeonUtilityBar from '../NeonUtilityBar/NeonUtilityBar';
import NeonMenu from '../NeonMenu/NeonMenu';

import { ROUTES, getFullRoute, buildAccountRoute } from '../../routing/routes';

const useStyles = makeStyles(() => ({
  root: {
    flexGrow: 1,
  },
}));

const NeonHeader = (props) => {
  const classes = useStyles();
  const { notifications, onShowNotifications } = props;

  const notificationsProps = (typeof onShowNotifications === 'function' ? {
    notifications, onShowNotifications,
  } : {});

  return (
    <div className={classes.root} data-selenium="neon-header">
      <NeonUtilityBar />
      <NeonMenu
        loginPath={getFullRoute(ROUTES.LOGIN)}
        logoutPath={getFullRoute(ROUTES.LOGOUT)}
        accountPath={buildAccountRoute()}
        {...notificationsProps}
      />
      <Divider />
    </div>
  );
};

NeonHeader.propTypes = {
  notifications: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      message: PropTypes.string.isRequired,
      dismissed: PropTypes.bool.isRequired,
    }),
  ),
  onShowNotifications: PropTypes.func,
};

NeonHeader.defaultProps = {
  notifications: [],
  onShowNotifications: null,
};

export default NeonHeader;
