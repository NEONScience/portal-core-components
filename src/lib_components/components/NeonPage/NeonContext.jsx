import React, {
  createContext,
  useContext,
  useReducer,
  useEffect,
} from 'react';
import PropTypes from 'prop-types';

import { of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';

import NeonGraphQL from '../NeonGraphQL/NeonGraphQL';
import sitesJSON from '../../static/sites/sites.json';
import statesJSON from '../../static/states/states.json';
import domainsJSON from '../../static/domains/domains.json';
import bundlesJSON from '../../static/bundles/bundles.json';

const FETCH_STATUS = {
  AWAITING_CALL: 'AWAITING_CALL',
  FETCHING: 'FETCHING',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
};

const DEFAULT_STATE = {
  data: {
    sites: {},
    states: statesJSON,
    domains: domainsJSON,
    bundles: bundlesJSON,
  },
  fetches: {
    sites: { status: FETCH_STATUS.AWAITING_CALL, error: null },
  },
};

/**
   CONTEXT
*/
const Context = createContext(DEFAULT_STATE);

/**
   HOOK
*/
const useNeonContextState = () => {
  const hookResponse = useContext(Context);
  if (hookResponse.length !== 2) {
    return [
      { ...DEFAULT_STATE },
      () => {},
    ];
  }
  return hookResponse;
};

/**
   Reducer
*/
const reducer = (state, action) => {
  const newState = { ...state };
  switch (action.type) {
    case 'fetchSitesCalled':
      newState.fetches.sites.status = FETCH_STATUS.FETCHING;
      return newState;
    case 'fetchSitesSucceeded':
      newState.fetches.sites.status = FETCH_STATUS.SUCCESS;
      newState.data.sites = action.sites;
      return newState;
    case 'fetchSitesFailed':
      newState.fetches.sites.status = FETCH_STATUS.ERROR;
      newState.fetches.sites.error = action.error;
      return newState;
    default:
      return state;
  }
};

/**
   Function to convert sites fetch response to the shape we expect
*/
const parseSitesFetchResponse = (sitesArray = []) => {
  if (!Array.isArray(sitesArray)) { return {}; }
  const sitesObj = {};
  sitesArray.forEach((site) => {
    sitesObj[site.siteCode] = {
      siteCode: site.siteCode || site.code,
      description: site.siteDescription || site.description,
      type: site.siteType || site.type,
      stateCode: site.stateCode,
      domainCode: site.domainCode,
      latitude: site.siteLatitude || site.latitude,
      longitude: site.siteLongitude || site.longitude,
      terrain: site.terrain || sitesJSON[site.siteCode].terrain,
    };
  });
  return sitesObj;
};

/**
   Context Provider
*/
const Provider = (props) => {
  const { children } = props;
  const [state, dispatch] = useReducer(reducer, { ...DEFAULT_STATE });

  // Subject and effect to perform and manage the sites GraphQL fetch
  const fetchAllSites$ = NeonGraphQL.getAllSites().pipe(
    map((response) => {
      if (response.response && response.response.data && response.response.data.sites) {
        const sites = parseSitesFetchResponse(response.response.data.sites);
        dispatch({ type: 'fetchSitesSucceeded', sites });
        return of(true);
      }
      dispatch({ type: 'fetchSitesFailed', error: 'malformed response' });
      return of(false);
    }),
    catchError((error) => {
      dispatch({ type: 'fetchSitesFailed', error: error.message });
      return of(false);
    }),
  );
  useEffect(() => {
    if (state.fetches.sites.status === FETCH_STATUS.AWAITING_CALL) {
      dispatch({ type: 'fetchSitesCalled' });
      fetchAllSites$.subscribe();
    }
  });

  /**
     Render
  */
  return (
    <Context.Provider value={[state]}>
      {children}
    </Context.Provider>
  );
};

Provider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.string,
    ])),
    PropTypes.node,
    PropTypes.string,
  ]).isRequired,
};

/**
   Export
*/
const NeonContext = { Provider, useNeonContextState };
export default NeonContext;
