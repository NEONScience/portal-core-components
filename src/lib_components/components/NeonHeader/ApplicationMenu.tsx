import React from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import IconButton from '@material-ui/core/IconButton';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Paper from '@material-ui/core/Paper';
import Popper from '@material-ui/core/Popper';
import Tooltip from '@material-ui/core/Tooltip';
import AppsIcon from '@material-ui/icons/Apps';
import Fade from '@material-ui/core/Fade';
import LaunchIcon from '@material-ui/icons/Launch';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import NeonContext from '../NeonContext/NeonContext';

// interface for user applications
interface UserApp {
  name: string;
  description: string;
  url: string;
}

// interface for the menu component props
interface MenuProps {
  apps: UserApp[];
}

// declare styles
const useStyles = makeStyles((theme: Theme) => createStyles({
  menuContainer: {
    zIndex: 1000, // be sure to display the menu over other elements
  },
  toolbarContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    boxShadow: '0px 1px 3px rgba(0, 0, 0, 0.45)',
    position: 'relative',
  },
  toolbarButtons: {
    display: 'flex',
    marginLeft: 'auto',
    marginRight: theme.spacing(2.5),
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
  },
  arrow: {
    color: theme.palette.common.black,
  },
  paper: {
    padding: theme.spacing(4),
    maxWidth: '500px', // limit width of menu
    marginTop: theme.spacing(1), // line top of menu up with divider
    overflowX: 'unset',
    overflowY: 'unset',
    '&::before': { // add tooltip like arrow to top of menu
      content: '""',
      position: 'absolute',
      marginRight: theme.spacing(4), // center arrow point beneath menu button
      top: 0,
      right: 0,
      width: theme.spacing(2), // width of arrow
      height: theme.spacing(2), // height of arrow
      backgroundColor: theme.palette.background.paper, // match paper background
      boxShadow: theme.shadows[2], // add arrow shadow
      transform: 'rotate(315deg)', // point arrow up toward menu button
      clipPath: 'polygon(-5px -5px, calc(100% + 5px) -5px, calc(100% + 5px) calc(100% + 5px))',
    },
  },
  card: {
    transition: '0.4s', // raised hover effect for Cards
    '&:hover': {
      transform: 'translateY(-2px)',
      boxShadow: '0px 1px 3px rgba(0, 0, 0, 0.45)',
    },
    cursor: 'pointer', // visually indicate Cards are links
    width: '100%', // ensure Cards are equal width
    border: 0,
  },
  cardContent: {
    textAlign: 'center',
  },
  gridItem: {
    display: 'flex', // so grid items stretch to equal height
  },
}));

const getApps = (): UserApp[] => {
  const [{ auth: authData }] = NeonContext.useNeonContextState();
  return authData?.userData?.data?.apps;
};

// define the menu component
const Menu = (props: MenuProps) => {
  const { apps } = props;
  const classes = useStyles();
  const [open, setOpen] = React.useState(false);
  const anchorRef = React.useRef<HTMLButtonElement>(null);

  // handle menu toggle
  const handleToggle = () => {
    setOpen((prevOpen) => !prevOpen);
  };

  // close the menu
  const handleClose = (event: React.MouseEvent<EventTarget>) => {
    if (anchorRef.current && anchorRef.current.contains(event.target as HTMLElement)) {
      return;
    }
    setOpen(false);
  };

  // open menu by tab key
  function handleMenuKeyDown(event: React.KeyboardEvent) {
    if (event.key === 'Tab') {
      event.preventDefault();
      setOpen(false);
    }
  }

  // handle a menu selection
  const handleMenuItemClick = (event: React.MouseEvent<EventTarget>, url: string) => {
    window.location.href = url;
  };

  // return focus to the button when toggling
  const prevOpen = React.useRef(open);
  React.useEffect(() => {
    if (prevOpen.current === true && open === false) {
      anchorRef.current!.focus();
    }
    prevOpen.current = open;
  }, [open]);

  return (
    <div className={classes.toolbarContainer}>
      <div className={classes.toolbarButtons}>
        <Tooltip
          title="Neon Applications"
          aria-label="Neon Applications"
          placement="left"
          TransitionComponent={Fade}
          TransitionProps={{ timeout: 200 }}
          arrow
        >
          <IconButton
            ref={anchorRef}
            style={{ color: 'black' }}
            aria-label="more"
            aria-controls={open ? 'neon-application-menu' : undefined}
            aria-haspopup="true"
            onClick={handleToggle}
            onKeyDown={handleMenuKeyDown}
          >
            <AppsIcon />
          </IconButton>
        </Tooltip>
        <Popper
          className={classes.menuContainer}
          open={open}
          anchorEl={anchorRef.current}
          role="presentation"
          transition
        >
          {({ TransitionProps, placement }) => (
            <Fade
              {...TransitionProps}
              style={{ transformOrigin: placement === 'bottom' ? 'center top' : 'center bottom' }}
              timeout={200}
            >
              <Paper elevation={3} className={classes.paper}>
                <ClickAwayListener onClickAway={handleClose}>
                  <Grid
                    container
                    spacing={4}
                    alignItems="stretch"
                  >
                    {apps.map((app: { name: string, description: string, url: string }) => (
                      <Grid
                        item
                        xs={(apps.length === 1) ? 12 : 6}
                        className={classes.gridItem}
                        key={app.name}
                      >
                        <Card
                          onClick={(event) => handleMenuItemClick(event, app.url)}
                          key={app.url}
                          className={classes.card}
                        >
                          <CardContent className={classes.cardContent}>
                            <LaunchIcon fontSize="large" />
                            <Typography variant="h6">{app.name}</Typography>
                            {app.description}
                          </CardContent>
                        </Card>
                      </Grid>
                    ))}
                  </Grid>
                </ClickAwayListener>
              </Paper>
            </Fade>
          )}
        </Popper>
      </div>
    </div>
  );
};

export default function ApplicationMenu() {
  const apps: UserApp[] = getApps();
  if (apps?.length > 0) {
    return (
      <Menu apps={apps} />
    );
  }
  return (null);
}
