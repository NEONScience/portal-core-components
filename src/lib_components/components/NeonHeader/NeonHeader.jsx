import React, {
  forwardRef,
  useState,
  useLayoutEffect,
} from 'react';
import PropTypes from 'prop-types';
import HTMLReactParser from 'html-react-parser';

import { makeStyles } from '@material-ui/core/styles';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import Skeleton from '@material-ui/lab/Skeleton';

import REMOTE_ASSETS from '../../remoteAssetsMap/remoteAssetsMap';
import Theme from '../Theme/Theme';
import NeonAuth, { NeonAuthType, NeonAuthDisplayType } from '../NeonAuth/NeonAuth';
import NeonEnvironment from '../NeonEnvironment/NeonEnvironment';
import NeonContext, { FETCH_STATUS } from '../NeonContext/NeonContext';

import NeonLegacyHeader from './NeonLegacyHeader';

const DRUPAL_HEADER_HTML_FALLBACK = require('../../remoteAssets/drupal-header.html');

const DRUPAL_HEADER_HTML = REMOTE_ASSETS.DRUPAL_HEADER_HTML.KEY;

const AUTH_ELEMENT_ID = 'header__authentication-ui';

const useStyles = makeStyles(theme => ({
  skeletonHeader: {
    boxShadow: '0px 1px 3px rgba(0, 0, 0, 0.25), 0px 1px 1px rgba(0, 0, 0, 0.25)',
  },
  coreAuthContainer: {
    // common styles
    textAlign: 'right',
    position: 'absolute',
    zIndex: 10,
    // viewport-specific styles
    [theme.breakpoints.up('lg')]: {
      padding: '0px',
      top: '-1px',
      right: '0px',
      '& :last-child': {
        borderRight: 'none',
        borderTopRightRadius: '0px',
        borderBottomRightRadius: '0px',
      },
      '& :first-child': {
        borderRight: 'none',
        borderTopLeftRadius: '0px',
      },
    },
    [theme.breakpoints.down('md')]: {
      padding: theme.spacing(1, 2),
      top: theme.spacing(1),
      right: theme.spacing(9),
    },
  },
  // These styles are gross. We need to rework the header coming from the Drupal site to make this
  // less necessary.
  unstickyHeader: {
    paddingTop: 'unset !important',
    '& > header': {
      position: 'unset !important',
      '& label[for="nav-trigger"]': {
        zIndex: '3 !important',
      },
      '& div.header__site-navigation': {
        zIndex: '2 !important',
      },
      [theme.breakpoints.down('sm')]: {
        '& .header__site-navigation': {
          display: 'none',
        },
        '& .nav-trigger:checked ~ .header__site-navigation': {
          display: 'block',
        },
      },
    },
  },
}));

const NeonHeader = forwardRef((props, headerRef) => {
  const {
    drupalCssLoaded,
    useCoreHeader,
    unstickyDrupalHeader,
  } = props;
  const classes = useStyles(Theme);
  const belowLg = useMediaQuery(Theme.breakpoints.down('md'));

  const [{
    isActive: neonContextIsActive,
    fetches: { [DRUPAL_HEADER_HTML]: headerFetch },
    html: { [DRUPAL_HEADER_HTML]: headerHTML },
    auth,
  }] = NeonContext.useNeonContextState();

  const [headerJsStatus, setHeaderJsStatus] = useState(FETCH_STATUS.AWAITING_CALL);
  const [headerRenderDelayed, setHeaderRenderDelayed] = useState(false);

  let renderMode = 'legacy';
  if (!useCoreHeader && neonContextIsActive) {
    switch (headerFetch.status) {
      case FETCH_STATUS.SUCCESS:
        renderMode = (headerHTML && drupalCssLoaded && headerRenderDelayed)
          ? 'drupal' : 'loading';
        break;
      case FETCH_STATUS.ERROR:
        renderMode = drupalCssLoaded ? 'drupal-fallback' : 'loading';
        break;
      default:
        renderMode = 'loading';
        break;
    }
  }

  // Load header.js only after initial delayed render of the drupal header is complete
  useLayoutEffect(() => {
    if (
      !renderMode.includes('drupal') || headerJsStatus !== FETCH_STATUS.AWAITING_CALL
        || !headerRenderDelayed || !drupalCssLoaded
    ) { return; }
    setHeaderJsStatus(FETCH_STATUS.FETCHING);
    const script = document.createElement('script');
    script.src = REMOTE_ASSETS.DRUPAL_HEADER_JS.url;
    script.onload = (() => {
      setHeaderJsStatus(FETCH_STATUS.SUCCESS);
    });
    script.onerror = (() => {
      setHeaderJsStatus(FETCH_STATUS.ERROR);
      // eslint-disable-next-line no-unused-expressions
      import('../../remoteAssets/drupal-header.js');
    });
    document.body.appendChild(script);
    // TODO: verify header.js loaded and if not load the fallback
  }, [headerJsStatus, drupalCssLoaded, headerRenderDelayed, setHeaderJsStatus, renderMode]);

  // Delay the rendering of the drupal header one render cycle to allow the CSS to propogate into
  // the environment. This prevents a "flash" of the unstyled menu in the drupal header on page load
  useLayoutEffect(() => {
    if (
      !useCoreHeader && neonContextIsActive
        && headerHTML && drupalCssLoaded && !headerRenderDelayed
    ) {
      const timeout = window.setTimeout(() => setHeaderRenderDelayed(true), 0);
      return () => window.clearTimeout(timeout);
    }
    return () => {};
  }, [
    neonContextIsActive,
    useCoreHeader,
    headerHTML,
    drupalCssLoaded,
    headerRenderDelayed,
  ]);

  // Render Loading
  if (renderMode === 'loading') {
    return (
      <header ref={headerRef} id="header" className={classes.skeletonHeader}>
        <Skeleton variant="rect" height={`${belowLg ? 60 : 125}px`} width="100%" />
      </header>
    );
  }

  // Render Drupal
  if (renderMode === 'drupal' || renderMode === 'drupal-fallback') {
    const injectAuth = !auth.useCore ? null : {
      replace: domNode => ((domNode.attribs || {}).id !== AUTH_ELEMENT_ID ? null : (
        <div id={AUTH_ELEMENT_ID} className={classes.coreAuthContainer}>
          <NeonAuth
            loginPath={NeonEnvironment.getFullAuthPath('login')}
            logoutPath={NeonEnvironment.getFullAuthPath('logout')}
            accountPath={NeonEnvironment.route.buildAccountRoute()}
            loginType={NeonAuthType.REDIRECT}
            logoutType={NeonAuthType.SILENT}
            displayType={NeonAuthDisplayType.MENU}
          />
        </div>
      )),
    };
    const html = renderMode === 'drupal' ? headerHTML : DRUPAL_HEADER_HTML_FALLBACK;
    return (
      <header
        ref={headerRef}
        id="header"
        className={unstickyDrupalHeader ? classes.unstickyHeader : null}
      >
        {HTMLReactParser(html, injectAuth)}
      </header>
    );
  }

  // Render Legacy
  return <NeonLegacyHeader {...props} ref={headerRef} />;
});

NeonHeader.propTypes = {
  ...NeonLegacyHeader.propTypes,
  drupalCssLoaded: PropTypes.bool,
  useCoreHeader: PropTypes.bool,
  unstickyDrupalHeader: PropTypes.bool,
};

NeonHeader.defaultProps = {
  ...NeonLegacyHeader.defaultProps,
  drupalCssLoaded: false,
  useCoreHeader: false,
  unstickyDrupalHeader: true,
};

export default NeonHeader;
