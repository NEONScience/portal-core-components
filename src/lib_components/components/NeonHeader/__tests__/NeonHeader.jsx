import React from 'react';
import renderer from 'react-test-renderer';

import '../../../../__mocks__/ajax';
import '../../../../__mocks__/NeonContext';
import NeonContext, { FETCH_STATUS } from '../../NeonContext/NeonContext';

import NeonHeader from '../NeonHeader';

import REMOTE_ASSETS from '../../../remoteAssetsMap/remoteAssetsMap';

jest.mock('../../../remoteAssets/drupal-header.html', () => (
  '<div>MOCK-FALLBACK-DRUPAL-HEADER-HTML</div>'
));
const DRUPAL_HEADER_HTML = REMOTE_ASSETS.DRUPAL_HEADER_HTML.KEY;

const defaultAuth = {
  useCore: false,
  isAuthenticated: false,
  isAuthWorking: false,
  isAuthWsConnected: false,
  userData: null,
};

describe('NeonHeader', () => {
  beforeEach(() => {
    NeonContext.useNeonContextState.mockReset();
  });
  test('renders with no props and inactive NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: false,
      fetches: { [DRUPAL_HEADER_HTML]: { status: null } },
      html: { [DRUPAL_HEADER_HTML]: null },
      auth: defaultAuth,
    }]);
    const tree = renderer
      .create(<NeonHeader />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with no props and active/fetching NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_HEADER_HTML]: { status: FETCH_STATUS.FETCHING } },
      html: { [DRUPAL_HEADER_HTML]: null },
      auth: defaultAuth,
    }]);
    const tree = renderer
      .create(<NeonHeader />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with no props and error NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_HEADER_HTML]: { status: FETCH_STATUS.ERROR } },
      html: { [DRUPAL_HEADER_HTML]: null },
      auth: defaultAuth,
    }]);
    const tree = renderer
      .create(<NeonHeader />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with no props and success NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_HEADER_HTML]: { status: FETCH_STATUS.SUCCESS } },
      html: { [DRUPAL_HEADER_HTML]: '<div>test drupal html</div>' },
      auth: defaultAuth,
    }]);
    const tree = renderer
      .create(<NeonHeader />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders fallback with drupalCSSLoaded prop and error NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_HEADER_HTML]: { status: FETCH_STATUS.ERROR } },
      html: { [DRUPAL_HEADER_HTML]: null },
      auth: defaultAuth,
    }]);
    const tree = renderer
      .create(<NeonHeader drupalCssLoaded />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with drupalCssLoaded prop and success NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_HEADER_HTML]: { status: FETCH_STATUS.SUCCESS } },
      html: { [DRUPAL_HEADER_HTML]: '<div>test drupal html</div>' },
      auth: defaultAuth,
    }]);
    const tree = renderer
      .create(<NeonHeader drupalCssLoaded unstickyDrupalHeader />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with drupalCssLoaded prop, success NeonContext state, and core auth', (done) => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_HEADER_HTML]: { status: FETCH_STATUS.SUCCESS } },
      html: { [DRUPAL_HEADER_HTML]: '<div>test drupal html</div>' },
      auth: { ...defaultAuth, useCore: true },
    }]);
    let tree;
    setTimeout(() => {
      renderer.act(() => {
        tree = renderer.create(<NeonHeader drupalCssLoaded />);
      });
      expect(tree.toJSON()).toMatchSnapshot();
      done();
    }, 0);
  });
});
