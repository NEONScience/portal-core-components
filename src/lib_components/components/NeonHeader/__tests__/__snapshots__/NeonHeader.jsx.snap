// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonHeader renders fallback with drupalCSSLoaded prop and error NeonContext state 1`] = `
<header
  className="makeStyles-unstickyHeader-28 makeStyles-headerContainer-30 makeStyles-headerContainerFallback-29"
  id="header"
>
  <div>
    MOCK-FALLBACK-DRUPAL-HEADER-HTML
  </div>
</header>
`;

exports[`NeonHeader renders with drupalCssLoaded prop and success NeonContext state 1`] = `
<header
  className="makeStyles-unstickyHeader-38 makeStyles-headerContainer-40"
  id="header"
>
  <div>
    test drupal html
  </div>
</header>
`;

exports[`NeonHeader renders with drupalCssLoaded prop, success NeonContext state, and core auth 1`] = `
<header
  className="makeStyles-unstickyHeader-48 makeStyles-headerContainer-50"
  id="header"
>
  <div>
    test drupal html
  </div>
</header>
`;

exports[`NeonHeader renders with no props and active/fetching NeonContext state 1`] = `
<header
  className="makeStyles-unstickyHeader-8 makeStyles-headerContainer-10 makeStyles-headerContainerFallback-9"
  id="header"
>
  <div>
    MOCK-FALLBACK-DRUPAL-HEADER-HTML
  </div>
</header>
`;

exports[`NeonHeader renders with no props and error NeonContext state 1`] = `
<header
  className="makeStyles-unstickyHeader-13 makeStyles-headerContainer-15 makeStyles-headerContainerFallback-14"
  id="header"
>
  <div>
    MOCK-FALLBACK-DRUPAL-HEADER-HTML
  </div>
</header>
`;

exports[`NeonHeader renders with no props and inactive NeonContext state 1`] = `
<header
  className="makeStyles-unstickyHeader-3 makeStyles-headerContainer-5 makeStyles-headerContainerFallback-4"
  id="header"
>
  <div>
    MOCK-FALLBACK-DRUPAL-HEADER-HTML
  </div>
</header>
`;

exports[`NeonHeader renders with no props and success NeonContext state 1`] = `
<header
  className="makeStyles-unstickyHeader-18 makeStyles-headerContainer-20 makeStyles-headerContainerFallback-19"
  id="header"
>
  <div>
    MOCK-FALLBACK-DRUPAL-HEADER-HTML
  </div>
</header>
`;
